### ISSUE TITLE
Failed calling webhook "webhook.tekton.dev": connection refused

### ISSUE THREAD

**Issue by maneilldev** [tektoncd/pipeline#2207]
# Expected Behavior
> kubectl apply -f example-git-resource.yaml
pipelineresource.tekton.dev/skaffold-git created

# Actual Behavior
> kubectl apply -f example-git-resource.yaml
Error from server (InternalError): error when creating "example-git-resource.yaml": Internal error occurred: failed calling webhook "webhook.tekton.dev": Post https://tekton-pipelines-webhook.tekton-pipelines.svc:443/?timeout=30s: dial tcp 10.99.167.76:443: connect: connection refused

# Steps to Reproduce the Problem
Follow the tekton example project in this repo on Docker Desktop on Mac running Kubernetes locally with the latest version of tekton

# Additional Info
kubectl apply --filename https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml
Client version: 0.8.0
Pipeline version: v0.11.0-rc1

This is the version that causes the error. If I moved back to v0.10, it's fine.
kubectl apply --filename https://storage.googleapis.com/tekton-releases/pipeline/previous/v0.10.0/release.yaml
kubectl apply -f example-git-resource.yaml
pipelineresource.tekton.dev/skaffold-git created

**kubectl version**
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:16:51Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:07:57Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}

**Comment by sbwsg** [comment]
I suggest trying this again.  We've released the beta version of Pipelines which includes a number of fixes since RC1 was released.

You should be able to run the same command to install the latest version:

```
kubectl apply --filename https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml
```

Also, just a data point, but I've recently installed Docker for Desktop on mac and the beta is working ok for me there.  At least Im not seeing any similar connection issues.

**Comment by sbwsg** [comment]
I'm going to close this issue since it's been almost a month with no further response. But please feel free to reopen if the issue persists or open a new issue if you've run in to different problems. Many thanks!

**Comment by lukehinds** [comment]
I am hitting this issue:

I used `examples/kind-with-registry.sh` to create a local registry declared within the `--config` passed to kind. 

I then used `ko` to load the current master branch of tektoncd/pipelines, but also used the latest release as well. 

I then loaded the a simple pipeline and task:

```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: sample-pipeline
spec:
  tasks:
    - name: echo-something
      taskRef:
        name: echo-hello-world
---
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: echo-hello-world
spec:
  steps:
    - name: echo
      image: ubuntu
      command:
        - echo
      args:
        - "hello world"
---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: sample-pipeline-run
spec:
  pipelineRef:
    name: sample-pipeline
```

This results in the following:

```
tkn pipelinerun logs -f sample-pipeline-run
E0529 12:21:57.833209   34220 reflector.go:123] pkg/mod/k8s.io/client-go@v0.16.5/tools/cache/reflector.go:96: Failed to list *v1alpha1.PipelineRun: conversion webhook for tekton.dev/v1beta1, Kind=PipelineRun failed: Post https://tekton-pipelines-webhook.tekton-pipelines.svc:443/?timeout=30s: dial tcp 10.99.190.15:443: connect: connection refused
E0529 12:21:59.882632   34220 reflector.go:123] pkg/mod/k8s.io/client-go@v0.16.5/tools/cache/reflector.go:96: Failed to list *v1alpha1.PipelineRun: conversion webhook for tekton.dev/v1beta1, Kind=PipelineRun failed: Post https://tekton-pipelines-webhook.tekton-pipelines.svc:443/?timeout=30s: dial tcp 10.99.190.15:443: connect: connection refused
```

I am curious if this might be an iptables config / issue.

I tried adding some end points:

```
cat <<EOF | kind create cluster --name "${KIND_CLUSTER_NAME}" --config=-
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  kubeadmConfigPatches:
  - |
    kind: InitConfiguration
    nodeRegistration:
      kubeletExtraArgs:
        node-labels: "ingress-ready=true"
  extraPortMappings:
  - containerPort: 443
    hostPort: 443
    protocol: TCP
containerdConfigPatches:
- |-
  [plugins."io.containerd.grpc.v1.cri".registry.mirrors."localhost:${reg_port}"]
    endpoint = ["http://${reg_ip}:${reg_port}"]
EOF
```

Unfortunately no luck.

**Comment by lukehinds** [comment]
@sbwsg could you reopen please?

**Comment by sbwsg** [comment]
I'm curious what the status of the webhook pod is.  You can check it with `kubectl get -n tekton-pipelines pod tekton-pipelines-webhook-xxxx` (those last `xxxx` will be a random string).

Something that at least works for me while developing locally is to delete the validating/mutating webhooks and then reapply the tekton yamls.  Something like the following:

```bash
kubectl delete validatingwebhookconfigurations.admissionregistration.k8s.io config.webhook.pipeline.tekton.dev
kubectl delete validatingwebhookconfigurations.admissionregistration.k8s.io validation.webhook.pipeline.tekton.dev
kubectl delete mutatingwebhookconfigurations.admissionregistration.k8s.io webhook.pipeline.tekton.dev
ko apply ./config # from the pipelines repo root
```

**Comment by nikhil-thomas** [comment]
@lukehinds are you still facing this issue.
if yes, could you check, `kubectl get crd pipelineruns.tekton.dev -o jsonpath='{.spec.conversion.webhook.clientConfig.service}'` verigy the service name and service namespace matches where the service is actually created

**Comment by bobcatfish** [comment]
@lukehinds let us know if you are still running into this, going to close for now!

**Comment by Fabian-K** [comment]
Hi, we are facing this issue when running some integration tests for our pipelines. Although the webhook pod is ready, creating tasks fails for some time with `connect: connection refused`. We are using kind version 0.10.0 for the cluster.

Here is a script that reproduces the issue. It 
(1) creates a new kind cluster
(2) waits for some core components to be ready
(3) installs tekton v0.23.0
(4) waits for the webhook pod to be ready and then 
(5) continuously tries to apply a sample task. This fails for some time until it eventually succeeds. 

I´m not sure if the root cause is somewhere inside tekton pipelines, kind or kubernetes itself 🙈  Do you have any suggestion how to investigate this further? 

Script:
```
echo "Creating cluster at $(date +"%T")"

kind create cluster

echo "Waiting for api-server pod at $(date +"%T")"

kubectl wait --namespace kube-system \
  --for=condition=ready pod \
  --selector=component=kube-apiserver \
  --timeout=120s

echo "Waiting for controller-manager pod at $(date +"%T")"

kubectl wait --namespace kube-system \
  --for=condition=ready pod \
  --selector=component=kube-controller-manager \
  --timeout=120s

echo "Waiting for core-dns pod at $(date +"%T")"

kubectl wait --namespace kube-system \
  --for=condition=ready pod \
  --selector=k8s-app=kube-dns \
  --timeout=120s

echo "Apply tekton pipelines at $(date +"%T")"

kubectl apply -f https://storage.googleapis.com/tekton-releases/pipeline/previous/v0.23.0/release.yaml

echo "Waiting for webhook pod at $(date +"%T")"

kubectl wait --namespace tekton-pipelines \
  --for=condition=ready pod \
  --selector=app.kubernetes.io/component=webhook \
  --timeout=120s

echo "Webhook ready at $(date +"%T")"

while :
do
  echo "Apply task at $(date +"%T")"
  kubectl apply -f task.yaml
	sleep 0.5s
done

```

Sample Execution:
```
Creating cluster at 11:46:15
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.20.2) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a nice day! 👋
Waiting for api-server pod at 11:46:53
pod/kube-apiserver-kind-control-plane condition met
Waiting for controller-manager pod at 11:46:58
pod/kube-controller-manager-kind-control-plane condition met
Waiting for core-dns pod at 11:48:09
pod/coredns-74ff55c5b-9kxwn condition met
pod/coredns-74ff55c5b-mtc7r condition met
Apply tekton pipelines at 11:48:09
namespace/tekton-pipelines created
podsecuritypolicy.policy/tekton-pipelines created
clusterrole.rbac.authorization.k8s.io/tekton-pipelines-controller-cluster-access created
clusterrole.rbac.authorization.k8s.io/tekton-pipelines-controller-tenant-access created
clusterrole.rbac.authorization.k8s.io/tekton-pipelines-webhook-cluster-access created
role.rbac.authorization.k8s.io/tekton-pipelines-controller created
role.rbac.authorization.k8s.io/tekton-pipelines-webhook created
role.rbac.authorization.k8s.io/tekton-pipelines-leader-election created
serviceaccount/tekton-pipelines-controller created
serviceaccount/tekton-pipelines-webhook created
clusterrolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller-cluster-access created
clusterrolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller-tenant-access created
clusterrolebinding.rbac.authorization.k8s.io/tekton-pipelines-webhook-cluster-access created
Warning: rbac.authorization.k8s.io/v1beta1 RoleBinding is deprecated in v1.17+, unavailable in v1.22+; use rbac.authorization.k8s.io/v1 RoleBinding
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller created
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-webhook created
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-controller-leaderelection created
rolebinding.rbac.authorization.k8s.io/tekton-pipelines-webhook-leaderelection created
customresourcedefinition.apiextensions.k8s.io/clustertasks.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/conditions.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/pipelines.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/pipelineruns.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/pipelineresources.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/runs.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/tasks.tekton.dev created
customresourcedefinition.apiextensions.k8s.io/taskruns.tekton.dev created
secret/webhook-certs created
validatingwebhookconfiguration.admissionregistration.k8s.io/validation.webhook.pipeline.tekton.dev created
mutatingwebhookconfiguration.admissionregistration.k8s.io/webhook.pipeline.tekton.dev created
validatingwebhookconfiguration.admissionregistration.k8s.io/config.webhook.pipeline.tekton.dev created
clusterrole.rbac.authorization.k8s.io/tekton-aggregate-edit created
clusterrole.rbac.authorization.k8s.io/tekton-aggregate-view created
configmap/config-artifact-bucket created
configmap/config-artifact-pvc created
configmap/config-defaults created
configmap/feature-flags created
configmap/config-leader-election created
configmap/config-logging created
configmap/config-observability created
configmap/config-registry-cert created
deployment.apps/tekton-pipelines-controller created
service/tekton-pipelines-controller created
horizontalpodautoscaler.autoscaling/tekton-pipelines-webhook created
deployment.apps/tekton-pipelines-webhook created
service/tekton-pipelines-webhook created
Waiting for webhook pod at 11:48:11
pod/tekton-pipelines-webhook-5bfbbd6475-5nx2m condition met
Webhook ready at 11:48:32
Apply task at 11:48:32
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:34
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:34
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:35
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:36
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:37
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:38
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:38
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:39
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": dial tcp 10.96.250.161:443: connect: connection refused
Apply task at 11:48:40
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": EOF
Apply task at 11:48:40
Error from server (InternalError): error when creating "task.yaml": Internal error occurred: failed calling webhook "webhook.pipeline.tekton.dev": Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/defaulting?timeout=10s": EOF
Apply task at 11:48:41
task.tekton.dev/example-task-name created
```

Thanks, Fabian

**Comment by Fabian-K** [comment]
In the meantime, I opened https://github.com/kubernetes/kubernetes/issues/102013 with respect to this issue. The first suggestion is to check if the webhook pod signals readiness too early.

https://github.com/tektoncd/pipeline/blob/55ae8563be29a76bb43b9577c8d6072ae7110d78/cmd/webhook/main.go#L218-L233

Looking at the webhook code, as far as I understand this signals readiness immediately. I´m wondering if that makes sense. As I understood from slack, the webhook is also responsible for generating the certificate required for the webhooks and this might take some time. Only once this is completed, k8s can call the webhook (because it needs the certificate?). Should not the webhook pod then wait for the certificate generation to be finished and only then signaling readiness (also taking failurePolicy: Fail into account)?

**Comment by junandaip** [comment]
> I'm curious what the status of the webhook pod is. You can check it with `kubectl get -n tekton-pipelines pod tekton-pipelines-webhook-xxxx` (those last `xxxx` will be a random string).
> 
> Something that at least works for me while developing locally is to delete the validating/mutating webhooks and then reapply the tekton yamls. Something like the following:
> 
> ```shell
> kubectl delete validatingwebhookconfigurations.admissionregistration.k8s.io config.webhook.pipeline.tekton.dev
> kubectl delete validatingwebhookconfigurations.admissionregistration.k8s.io validation.webhook.pipeline.tekton.dev
> kubectl delete mutatingwebhookconfigurations.admissionregistration.k8s.io webhook.pipeline.tekton.dev
> ko apply ./config # from the pipelines repo root
> ```

still works 2022

**Comment by drriguz** [comment]
Got the same error somehow in my local minikube environment:
```
kubectl apply -f hello-world-task.yaml
Error from server: error when retrieving current configuration of:
Resource: "tekton.dev/v1, Resource=tasks", GroupVersionKind: "tekton.dev/v1, Kind=Task"
Name: "hello-world-task", Namespace: "default"
from server for: "hello-world-task.yaml": conversion webhook for tekton.dev/v1beta1, Kind=Task failed: Post "https://tekton-pipelines-webhook.tekton-pipelines.svc:443/resource-conversion?timeout=30s": dial tcp 10.105.168.23:443: connect: connection refused
```

Restart minikube solved the problem, but I don't know the reason.

**Comment by kinget007** [comment]
Deployment of tekton-pipelines-webhook and  tekton-pipelines-controller: 
spec:
      hostNetwork: true

**Comment by terryzwt** [comment]
> Deployment of tekton-pipelines-webhook and tekton-pipelines-controller: spec: hostNetwork: true

Not work in my case.

**Comment by sidofdountio** [comment]
I have the same error: 
![error-kubectl](https://github.com/tektoncd/pipeline/assets/69512885/6b32abd0-f32c-42e1-adc9-29889adfc304)
To solve it i just restart minikube:
`minikube start`
